### PR TITLE
cmake: build boost as an external project

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

**Comment by tchaikov** [comment]
@liewegas i think in this way

- the compilation will speed up a little bit without pulling the whole boost repo and its submodules.
- and we don't need to rerun `src/boost/bootstrap.sh` and `b2` every time `cmake` is launched

i will update `make-dist` to include the boost tarball if this looks fine to you.

**Comment by tchaikov** [comment]
retest this please.

**Comment by Liuchang0812** [comment]
Hi, Kefu, are you working on this? It would be great for me.... terrible network in the China...

**Comment by tchaikov** [comment]
@Liuchang0812 i don't follow you. this PR sent by me, so presumably i am working on this?

**Comment by Liuchang0812** [comment]
sorry for confusing you. just hope it could be merged as soon as possible.

**Comment by tchaikov** [comment]
it's RFC, and pending on some early review.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -300,6 +300,7 @@ endif()
 # tcmalloc heap profiler
 set(heap_profiler_files ${TCMALLOC_srcs})
 add_library(heap_profiler_objs OBJECT ${heap_profiler_files})
+add_dependencies(heap_profiler_objs Boost::boost)
```
Comment (cbodley) [inline src/CMakeLists.txt]: these deps are how we trigger the external project, and guarantee that the boost headers are available?

if we add a new target that doesn't add this dependency, does it fail somehow? or would it silently pull in the boost system headers?

this seems error-prone, but i'm not sure how to reconcile this with the need for `include_directories(BEFORE SYSTEM ${Boost_INCLUDE_DIRS})` to override the system headers

Comment (tchaikov) [inline src/CMakeLists.txt]: > these deps are how we trigger the external project, and guarantee that the boost headers are available?

yes.

> if we add a new target that doesn't add this dependency, does it fail somehow? or would it silently pull in the boost system headers?

yes. no, unless that target depends on some other target which depends on `Boost::boost`.

> this seems error-prone, but i'm not sure how to reconcile this with the need for `include_directories(BEFORE SYSTEM ${Boost_INCLUDE_DIRS})` to override the system headers

i feel the same. i was evaluating some alternatives. but none of them are feasible or elegant enough. for example, to preprocess CMakeLists.txt using cmake and add `add_dependencies(${THE_TARGET} Boost::boost)` for every target.

Comment (cbodley) [inline src/CMakeLists.txt]: > for example, to preprocess CMakeLists.txt using cmake and add add_dependencies(${THE_TARGET} Boost::boost) for every target.

we might consider something [hacky like this](https://cmake.org/pipermail/cmake/2014-August/058418.html), which overrides `add_library()` and `add_executable()` to inject a global dependency

Comment (tchaikov) [inline src/CMakeLists.txt]: that's .... hacky! i will give it a shot. thanks for the link @cbodley !

**Comment by tchaikov** [comment]
i am removing the commit to remove the boost submodule. will try to get this one merged first, and have that commit in a follow-up PR.

**Comment by tchaikov** [comment]
@cbodley fixed and repushed, could you take a look again?

File: CMakeLists.txt [hunk 2]
```diff
@@ -534,78 +539,26 @@ if(WITH_RADOSGW_BEAST_FRONTEND)
...
-if (WITH_SYSTEM_BOOST)
+set(Boost_USE_MULTITHREADED ON)
+# require minimally the bundled version
+if(WITH_SYSTEM_BOOST)
   if(ENABLE_SHARED)
     set(Boost_USE_STATIC_LIBS OFF)
   else()
     set(Boost_USE_STATIC_LIBS ON)
   endif()
+  find_package(Boost 1.61 COMPONENTS ${BOOST_COMPONENTS} REQUIRED)
```
Comment (cbodley) [inline CMakeLists.txt]: is there a reason this is staying at 1.61, when `BuildBoost.cmake` and the submodule are at 1.63?

**Comment by cbodley** [comment]
> i am removing the commit to remove the boost submodule. will try to get this one merged first, and have that commit in a follow-up PR.

if this doesn't remove the submodule, but adds an external project that fetches the boost tarball separately, is there a net benefit to merging this as it is?

File: CMakeLists.txt (hunk 2 above)
Comment (tchaikov) [inline CMakeLists.txt]: yes, if user happens to have 1.61 in his/her system, but he/she does not want to rebuild boost. then he/she will benefit from this.

**Comment by tchaikov** [comment]
> if this doesn't remove the submodule, but adds an external project that fetches the boost tarball separately, is there a net benefit to merging this as it is?

@cbodley,

- no more mysterious cmake warnings from `FindBoost`,
- no more `./b2` calls when cmake is launched. 
- and some developers with slow internet access would be happier if the "src/boost" is optional. believe it or not, it took me more than 1 hour to pull the whole boost repo from github.

**Comment by Liuchang0812** [comment]
`it took me more than 1 hour to pull the whole boost repo from github for the first time.`

can't agree more....

File: CMakeLists.txt (hunk 2 above)
Comment (cbodley) [inline CMakeLists.txt]: ah right, this is the `WITH_SYSTEM_BOOST` part. what about the `build_boost(1.61` part below?

File: CMakeLists.txt [hunk 3]
```diff
@@ -22,7 +22,12 @@ endif()
 if(POLICY CMP0065)
   cmake_policy(SET CMP0065 NEW)
 endif()
-
+if(POLICY CMP0051)
+  # cmake 3.1 and higher include generator expressions in SOURCES property,
+  # but i don't want to bother with stripping off the TARGET_OBJECTS elements
+  # from it. so let's stick with the old behavior now.
+  cmake_policy(SET CMP0051 OLD)
+endif()
```
Comment (cbodley) [inline CMakeLists.txt]: okay, it took me a while to realize this was related to `get_target_property(sources ${target} SOURCES)` in `maybe_add_boost_dep()`. could you mention `maybe_add_boost_dep()` in this comment?

File: CMakeLists.txt (hunk 2 above)
Comment (tchaikov) [inline CMakeLists.txt]: yeah, i was trying to map the logic of `FindBoost` to `BuildBoost`, and wanted to make `BuildBoost` more generic in hope that other projects are able to leverage this cmake module  with less work. if it seems useless and over-engineering, i am happy to remove this cruft.

File: CMakeLists.txt (hunk 3 above)
Comment (adamemerson) [inline CMakeLists.txt]: Even if we keep the submodule it's worth having boost build when we're building and not configuing.

**Comment by tchaikov** [comment]
changelog

- amend the comment for the reason why we have `cmake_policy(SET CMP0051 OLD)`

File: CMakeLists.txt (hunk 2 above)
Comment (tchaikov) [inline CMakeLists.txt]: sure, updated!

**Comment by tchaikov** [comment]
@cbodley thanks for your pointers and the review, will merge it once shaman is happy: https://shaman.ceph.com/builds/ceph/wip-15376-kefu/

**Comment by Liuchang0812** [comment]
Hi, Kefu, I tested this patch today. It seems that we don't remove boost submodule.  we will pull whole boost repos when we execute `./do_cmake.sh`.  Need we remove boost submodule or update `do_cmake.sh` script?

```
➜  test-cmake git clone https://github.com/ceph/ceph.git
Cloning into 'ceph'...
remote: Counting objects: 515228, done.
remote: Compressing objects: 100% (26/26), done.
remote: Total 515228 (delta 12), reused 9 (delta 1), pack-reused 515201
Receiving objects: 100% (515228/515228), 215.12 MiB | 352.00 KiB/s, done.
Resolving deltas: 100% (406117/406117), done.
Checking connectivity... done.
➜  test-cmake cd ceph 
➜  ceph git:(master) ./do_cmake.sh 
+ git submodule update --init --recursive
Submodule 'ceph-erasure-code-corpus' (https://github.com/ceph/ceph-erasure-code-corpus.git) registered for path 'ceph-erasure-code-corpus'
Submodule 'ceph-object-corpus' (https://github.com/ceph/ceph-object-corpus.git) registered for path 'ceph-object-corpus'
Submodule 'src/Beast' (https://github.com/ceph/Beast.git) registered for path 'src/Beast'
Submodule 'src/blkin' (https://github.com/ceph/blkin) registered for path 'src/blkin'
Submodule 'src/boost' (https://github.com/boostorg/boost.git) registered for path 'src/boost'
Submodule 'src/civetweb' (https://github.com/ceph/civetweb) registered for path 'src/civetweb'
Submodule 'src/crypto/isa-l/isa-l_crypto' (https://github.com/01org/isa-l_crypto) registered for path 'src/crypto/isa-l/isa-l_crypto'
Submodule 'src/dpdk' (https://github.com/ceph/dpdk) registered for path 'src/dpdk'
Submodule 'src/erasure-code/jerasure/gf-complete' (https://github.com/ceph/gf-complete.git) registered for path 'src/erasure-code/jerasure/gf-complete'
Submodule 'src/erasure-code/jerasure/jerasure' (https://github.com/ceph/jerasure.git) registered for path 'src/erasure-code/jerasure/jerasure'
Submodule 'src/googletest' (https://github.com/ceph/googletest) registered for path 'src/googletest'
Submodule 'src/isa-l' (https://github.com/ceph/isa-l) registered for path 'src/isa-l'
Submodule 'src/lua' (https://github.com/ceph/lua.git) registered for path 'src/lua'
Submodule 'src/rapidjson' (https://github.com/ceph/rapidjson) registered for path 'src/rapidjson'
Submodule 'src/rocksdb' (https://github.com/ceph/rocksdb) registered for path 'src/rocksdb'
Submodule 'src/spdk' (https://github.com/ceph/spdk.git) registered for path 'src/spdk'
Submodule 'src/xxHash' (https://github.com/ceph/xxHash.git) registered for path 'src/xxHash'
Submodule 'src/zstd' (https://github.com/facebook/zstd) registered for path 'src/zstd'
Cloning into 'ceph-erasure-code-corpus'...
remote: Counting objects: 3362, done.
remote: Compressing objects: 100% (60/60), done.
```

**Comment by tchaikov** [comment]
@Liuchang0812 see https://github.com/ceph/ceph/pull/15376#issuecomment-308316730 and https://github.com/ceph/ceph/pull/15376#issuecomment-305168444

**Pull request merged**